### ISSUE TITLE
question about `examples/ibex`

### ISSUE THREAD

**Issue by hchengv** [IAIK/coco-alma#13]
Hello, 

I'm trying to run the `examples/ibex` and using my own program which is:
```
.org 0x80
nop
xor x1,  x1, x9
and x0,  x0, x0
xor x17, x1, x9
nop
nop
nop
...
```
and the label file is:
```
# registers: x1, x9, x17
u_core.register_file_i.rf_reg_reg[32]  = secret 0
u_core.register_file_i.rf_reg_reg[288] = static_random
u_core.register_file_i.rf_reg_reg[544] = secret 0
```

I execute the verification script using the `--mode stable` and the `--include-hamming` options (plus `--cycles -1`, `--rst-name rst_sys_n`, `--rst-cycles 2`, `--rst-phase 0`), and I expect it should report a transition leakage (because I overwrite the register `x17` which contains a share by another share). But the verification script doesn't report any leakages, therefore I'd like to ask: is there somewhere I made a mistake (about the usage of this tool) or somewhere I just didn't understand correctly?

Many thanks for your help!

**Comment by barbara-gigerl** [comment]
Hi,
in order to capture transitions, you must use the transient verification mode (`--mode transient`)
BR

**Comment by hchengv** [comment]
Hi,

Thanks for your fast answer. 
According to my experiments, using `--mode transient --glitch-behavior loose` cannot detect the leakage I expect, and only using `--mode transient --glitch-behavior strict` reports the leakage. Is this supposed to be correct?

**Comment by hchengv** [comment]
Hi,

Sorry, it's still not very clear to me. Because the leakage in my program seems to be obvious, so I guess just using `--mode transient --glitch-behavior loose` options (i.e., do not need `--glitch-behavior strict`) should report this leakage, right? 

Thanks!

**Comment by vedadux** [comment]
Hi,

sorry for not responding earlier since I didn't see the issue notification. 
I also believe this should show up as leakage with either `--mode transient --glitch-behavior strict` or with `--mode stable --include-hamming`. In general, it should also show up with `--mode transient --glitch-behavior loose`, since that should imply `--include-hamming`. I will investigate and get back to you.

Vedad

**Comment by vedadux** [comment]
I have recreated your test, and the intern hamming tracking works as expected, and actually indicates leakage, i.e., I get the following debug output:
```
hamming dff u_core.register_file_i.rf_reg_reg[544][0] : PropVarSet(id=12, vars={}, ones={0, 2})
```
where 0 and 2 are the indices of the two shares s_0 and s_1, and the tool is essentially detecting that the hamming distance in s_0 ^ s_1. However, this is somehow missed in the check function.
I will update you as soon as I figure it out fully, and create a patch.

**Comment by hchengv** [comment]
Hi Vedad,

Many thanks for your help. That's great the issue doesn't seem to be so difficult to fix.

Best regards,
Hao

**Comment by vedadux** [comment]
Hi Hao, 

I have now identified the issue with the checking. The bug was introduced when the old per-secret checking methodology was merged back into the master. It was revived from old code and assumes either 'stable' or 'transient strict' modes are active, which leads to hamming leakage being ignored when 'transient loose' or simply 'stable include-hamming' is used. If you use `--checking-mode per-location` it should work as desired and find the leak, although it is reportedly slower for higher order checking. I am now working on a fix, but it needs some re-structuring again, which could take some time.

Best regards,
Vedad
